src/io/platform/f3xx/f302x8/SPIf302x8.cpp

```diff
@@ -0,0 +1,359 @@
+//
+// Created by matth on 10/2/2021.
+//
+
+#include <EVT/io/pin.hpp>
+#include <EVT/io/platform/f3xx/f302x8/SPIf302x8.hpp>
+
+namespace EVT::core::IO {
+/**
+     * gets the corresponding SPI port from the mosi pin
+     * @param mosiPin the pin to use for the mosi signal
+     * @return the SPI port to use, 0 if invalid
+     */
+uint8_t SPIf302x8::getMOSIPortID(Pin mosiPin) {
+    switch (mosiPin) {
+    case Pin::PA_11:
+    case Pin::PB_15:
+        return 2;
+    case Pin::PB_5:
+    case Pin::PC_12:
+        return 3;
+    default:
+        return 0;
+    }
+}
+
+/**
+     * gets the corresponding SPI port from the miso pin
+     * @param misoPin the pin to use for the miso signal
+     * @return the SPI port to use, 0 if invalid
+     */
+uint8_t SPIf302x8::getMISOPortID(Pin misoPin) {
+    switch (misoPin) {
+    case Pin::PA_10:
+    case Pin::PB_14:
+        return 2;
+    case Pin::PB_4:
+    case Pin::PC_11:
+        return 3;
+    default:
+        return 0;
+    }
+}
+
+/**
+     * gets the corresponding SPI port from the sck pin
+     * @param sckPin the pin to use for the sck signal
+     * @return the SPI port to use, 0 if invalid
+     */
+uint8_t SPIf302x8::getSCKPortID(Pin sckPin) {
+    switch (sckPin) {
+    case Pin::PA_10:
+    case Pin::PF_1:
+        return 2;
+    case Pin::PB_3:
+    case Pin::PC_10:
+        return 3;
+    default:
+        return 0;
+    }
+}
+
+/**
+     * Constructs a send and receive SPI object
+     * @param CSPins an array of chip select pins for selecting which device to communicate with.
+     * @param pinLength the number of pins in the chip select array
+     * @param sckPin the pin for the clk line
+     * @param mosiPin the mosi pin for sending data
+     * @param misoPin the miso pin for receiving data
+     */
+SPIf302x8::SPIf302x8(GPIO** CSPins, uint8_t pinLength, Pin sckPin, Pin mosiPin, Pin misoPin) : SPI(CSPins, pinLength, sckPin, mosiPin, misoPin) {
+    uint8_t mosiPort = getMOSIPortID(mosiPin);
+    uint8_t misoPort = getMISOPortID(misoPin);
+    uint8_t sckPort = getSCKPortID(sckPin);
+    GPIO_InitTypeDef GPIOInit = {0};
```

Comment:
For the GPIO setup. We now have have a method for abstracting the GPIO init logic. See [this function](https://github.com/RIT-EVT/EVT-core/blob/main/include/EVT/io/platform/f3xx/f302x8/GPIOf302x8.hpp#L69)

Reply:
Will this function take into account the specific pins that have SPI hardware and that all the pins are on the same SPI port?

Reply:
It doesn't it really replaces lines 96-129.

Reply:
oh, ok